File: PYTHON/Cryptor.py
from MyRandom import MyRandom

class Cryptor(object):
    _KEY="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!?"
    _KEY_LENGTH = 64
    SALT_LENGTH = 16
    MULTIPLY = 1.2
    DISORDER_RATE = 6
    def __init__(self):
        self.r=MyRandom("a")
    def __randomString(self,length):
        retu=""
        for i in range(0,length):
            retu += Cryptor._KEY[self.r.generate() % Cryptor._KEY_LENGTH]
        return retu
    def __makeDisorderMoves(length,key):
        loopCount = Cryptor.DISORDER_RATE * length * 2
        arr = [0]*loopCount
        ran = MyRandom(key)
        for i in range(0,loopCount):
            arr[i] = ran.generate() % length
        return arr
    def __makeDisorderArray(length,disorderMoves):
        loopCount = Cryptor.DISORDER_RATE * length * 2
        ret = [i for i in range(0,length)]
        for i in range(0,loopCount,2):
            ret[disorderMoves[i]],ret[disorderMoves[i + 1]] = ret[disorderMoves[i + 1]],ret[disorderMoves[i]]
        return ret
    def __bitSet(input,position,toWhat):
        pow2 = 2**position
        if (((input & pow2)!=0) == toWhat):
            return input
        if (toWhat):
            return input | pow2
        else:
            return input & (63 - pow2)
    def __bitGet(input,position):
        pow2 = 2**position
        return (input & pow2)!=0
    def __getPossition(input):
        return Cryptor._KEY.find(input)
    def encrypt(self,input,key):
        input = input.encode("utf8")
        retu = ""
        pointer = 0
        length = int(Cryptor.MULTIPLY * (len(input)+1) * 8)
        length = length + 6 - length % 6
        salt = self.__randomString(Cryptor.SALT_LENGTH)
        disorderMoves = Cryptor.__makeDisorderMoves(length, key + salt)
        disorderArray = Cryptor.__makeDisorderArray(length, disorderMoves)
        chars = [0] * int(length / 6)
        for i in range(0,len(input)):
            for p in range(0,8):
                j = 2**p
                chars[int(disorderArray[pointer] / 6)] = Cryptor.__bitSet(chars[int(disorderArray[pointer] / 6)], disorderArray[pointer] % 6, (input[i] & j)!=0)
                pointer+=1
        for p in range(0,8):
            j = 2**p
            chars[int(disorderArray[pointer] / 6)] = Cryptor.__bitSet(chars[int(disorderArray[pointer] / 6)], disorderArray[pointer] % 6, 0)
            pointer+=1
        while pointer < length:
            chars[int(disorderArray[pointer] / 6)] = Cryptor.__bitSet(chars[int(disorderArray[pointer] / 6)], disorderArray[pointer] % 6, (self.r.generate() % 2)!=0)
            pointer+=1
        for i in range(0,int(length/6)):
            retu += Cryptor._KEY[chars[i]]
        return retu + salt
    def decrypt(self,input,key):
        retu = bytearray()
        salt = input[len(input) - Cryptor.SALT_LENGTH:]
        input = input[:len(input) - Cryptor.SALT_LENGTH]
        length = len(input) * 6
        disorderMoves = Cryptor.__makeDisorderMoves(length, key + salt)
        disorderArray = Cryptor.__makeDisorderArray(length, disorderMoves)
        chars = []
        for i in range(0,int(length/6)):
            chars.append(Cryptor.__getPossition(input[i]))
        pointer = 0
        while pointer < length:
            temp = 0
            for j in range (0,8):
                temp = Cryptor.__bitSet(temp, j, Cryptor.__bitGet(chars[int(disorderArray[pointer] / 6)], disorderArray[pointer] % 6))
                pointer+=1
                if pointer == length:
                    break
            if temp == 0:
                break
            retu.append(temp)
        return retu.decode('utf8')

File: JAVA/Cryptor.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package mpd;

import java.io.UnsupportedEncodingException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author mohalisad
 */
public class Cryptor {
    private final static String _KEY = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!?";
    private final static int _KEY_LENGTH = 64;
    private final static int SALT_LENGTH = 16;
    private final static double MULTIPLY = 1.2;
    private final static int DISORDER_RATE = 6;
    private MyRandom r=new MyRandom (Date.now());
    private String randomString(int length)
    {
        String retu="";
        for (int i = 0; i < length; i++)
        {
            retu += _KEY.charAt(r.generate() % _KEY_LENGTH);
        }
        return retu;
    }
    private static int[] makeDisorderMoves(int length, String key)
    {
        int loopCount = DISORDER_RATE * length * 2;
        int[] arr = new int[loopCount];
        MyRandom ran=new MyRandom(key);
        for (int i = 0; i < loopCount; i++)
        {
            arr[i] = ran.generate() % length;
        }
        return arr;
    }
    private static int[] makeDisorderArray(int length, int[] disorderMoves)
    {
        int c;
        int loopCount = DISORDER_RATE * length * 2;
        int[] ret = new int[length];
        for (int i = 0; i < length; i++)
        {
            ret[i] = i;
        }
        for (int i = 0; i < loopCount; i += 2)
        {
            c = ret[disorderMoves[i]];
            ret[disorderMoves[i]] = ret[disorderMoves[i + 1]];
            ret[disorderMoves[i + 1]] = c;
        }
        return ret;
    }
    private static int bitSet(int input, int position, boolean toWhat)
    {
        int pow2 = 1;
        for (int i = 0; i < position; i++) pow2 *= 2;
        if (((input & pow2)!=0) == toWhat) return input;
        if (toWhat)
        {
            return input | pow2;
        }
        else
        {
            return input & (63 - pow2);
        }
    }
    private static boolean bitGet(int input, int position)
    {
        int pow2 = 1;
        for (int i = 0; i < position; i++) pow2 *= 2;
        return (input & pow2)!=0;
    }
    private static int getPossition(char input)
    {
        return _KEY.indexOf(input);
    }
    public String encrypt(String _input, String key)
    {
        String retu="", salt;
        int[] disorderArray,disorderMoves,chars;
        byte[] input = _input.getBytes(StandardCharsets.UTF_8);
        int pointer = 0, length = (int)(MULTIPLY * (input.length+1) * 8);
        length = length + 6 - length % 6;
        salt = randomString(SALT_LENGTH);
        disorderMoves = makeDisorderMoves(length, key + salt);
        disorderArray = makeDisorderArray(length, disorderMoves);
        chars = new int[length / 6];
        for (int i = 0; i < length / 6; i++)
        {
            chars[i] = 0;
        }
        for (int i = 0; i < input.length; i++)
        {
            for (int j = 1; j < 256; j = j * 2)
            {
                chars[disorderArray[pointer] / 6] = bitSet(chars[disorderArray[pointer] / 6], disorderArray[pointer] % 6, (input[i] & j)!=0);
                pointer++;
            }
        }
        for (int j = 1; j < 256; j = j * 2)
        {
            chars[disorderArray[pointer] / 6] = bitSet(chars[disorderArray[pointer] / 6], disorderArray[pointer] % 6, false);
            pointer++;
        }
        for (; pointer < length; pointer++)
        {
            chars[disorderArray[pointer] / 6] = bitSet(chars[disorderArray[pointer] / 6], disorderArray[pointer] % 6, (r.generate() % 2)!=0);
        }
        for (int i = 0; i < length / 6; i++)
        {
            retu += _KEY.charAt(chars[i]);
        }
        return retu + salt;
    }
    public String decrypt(String input, String key)
    {
        int[] disorderArray,disorderMoves,chars;
        ArrayList<Byte> retchars = new ArrayList<Byte>();
        byte[] _retchars;
        Byte temp;
        int length;
        String salt = input.substring(input.length() - SALT_LENGTH, input.length());
        input = input.substring(0, input.length() - SALT_LENGTH);
        length = input.length() * 6;
        disorderMoves = makeDisorderMoves(length, key + salt);
        disorderArray = makeDisorderArray(length, disorderMoves);
        chars = new int[length / 6];
        for (int i = 0; i < length / 6; i++)
        {
            chars[i] = getPossition(input.charAt(i));
        }
        int pointer = 0;
        while (pointer < length)
        {
            temp = 0;
            for (int j = 0; j < 8; j++)
            {
                temp = (byte)bitSet(temp, j, bitGet(chars[disorderArray[pointer] / 6], disorderArray[pointer] % 6));
                pointer++;
                if (pointer == length) break;
            }
            if (temp == 0) break;
            retchars.add(temp);
        }
        _retchars=new byte[retchars.size()];
        for(int i=0;i<retchars.size();i++){
            _retchars[i]=retchars.get(i).byteValue();
        }
        try {
            return new String(_retchars,"UTF-8");
        } catch (Exception ex) {
            return "";
        }
    }
}
